lib/Versions/VersionsBackend.php

```diff
@@ -32,66 +35,118 @@
-		$this->mountProvider = $mountProvider;
-		$this->timeFactory = $timeFactory;
-		$this->logger = $logger;
+class VersionsBackend implements IVersionBackend, INameableVersionBackend, IDeletableVersionBackend, INeedSyncVersionBackend {
+	public function __construct(
+		private IRootFolder $rootFolder,
+		private Folder $appFolder,
+		private MountProvider $mountProvider,
+		private ITimeFactory $timeFactory,
+		private LoggerInterface $logger,
+		private GroupVersionsMapper $groupVersionsMapper,
+		private IMimeTypeLoader $mimeTypeLoader,
+	) {
 	}
 
 	public function useBackendForStorage(IStorage $storage): bool {
 		return true;
 	}
 
-	public function getVersionsForFile(IUser $user, FileInfo $file): array {
-		$mount = $file->getMountPoint();
-		if ($mount instanceof GroupMountPoint) {
-			try {
-				$folderId = $mount->getFolderId();
-				/** @var Folder $versionsFolder */
-				$versionsFolder = $this->getVersionsFolder($mount->getFolderId())->get((string)$file->getId());
-				$versions = array_map(function (Node $versionFile) use ($file, $user, $folderId): GroupVersion {
-					if ($versionFile instanceof Folder) {
-						$this->logger->error('Found an unexpected subfolder inside the groupfolder version folder.');
-					}
-					return new GroupVersion(
-						(int)$versionFile->getName(),
-						(int)$versionFile->getName(),
-						$file->getName(),
-						$versionFile->getSize(),
-						$versionFile->getMimetype(),
-						$versionFile->getPath(),
-						$file,
-						$this,
-						$user,
-						$versionFile,
-						$folderId
-					);
-				}, $versionsFolder->getDirectoryListing());
-				usort($versions, function (GroupVersion $v1, GroupVersion $v2): int {
-					return $v2->getTimestamp() <=> $v1->getTimestamp();
-				});
+	public function getVersionsForFile(IUser $user, FileInfo $fileInfo): array {
+		$mount = $fileInfo->getMountPoint();
+		if (!($mount instanceof GroupMountPoint)) {
+			return [];
+		}
+
+		try {
+			$folderId = $mount->getFolderId();
+			/** @var Folder $versionsFolder */
+			$versionsFolder = $this->getVersionsFolder($mount->getFolderId())->get((string)$fileInfo->getId());
+
+			$userFolder = $this->rootFolder->getUserFolder($user->getUID());
+			$nodes = $userFolder->getById($fileInfo->getId());
+			$file = array_pop($nodes);
+
+			$versions = $this->getVersionsForFileFromDB($fileInfo, $user, $folderId);
+			if (count($versions) > 0) {
 				return $versions;
-			} catch (NotFoundException $e) {
-				return [];
 			}
-		} else {
+
+			// Insert the entry in the DB for the current version.
+			$versionEntity = new GroupVersionEntity();
+			$versionEntity->setFileId($file->getId());
+			$versionEntity->setTimestamp($file->getMTime());
+			$versionEntity->setSize($file->getSize());
+			$versionEntity->setMimetype($this->mimeTypeLoader->getId($file->getMimetype()));
+			$versionEntity->setDecodedMetadata([]);
+			$this->groupVersionsMapper->insert($versionEntity);
+
+			// Insert entries in the DB for existing versions.
```

Comment:
I’m lost, why is that needed outside of migrations?

Reply:
Ok, got it, outside of migration the function called before will find versions in DB and this will not run.